src/EditorFeatures/Core.Wpf/InlineHints/InlineHintsTagger.cs

```diff
@@ -126,86 +106,60 @@ private TextFormattingRunProperties Format
             }
         }
 
-        private void InvalidateCache()
-        {
-            _taggerProvider.ThreadingContext.ThrowIfNotOnUIThread();
-            _cacheSnapshot = null;
-            _cache.Clear();
-        }
-
-        IEnumerable<ITagSpan<IntraTextAdornmentTag>> ITagger<IntraTextAdornmentTag>.GetTags(NormalizedSnapshotSpanCollection spans)
-            => GetTags(spans);
-
-        public IReadOnlyList<TagSpan<IntraTextAdornmentTag>> GetTags(NormalizedSnapshotSpanCollection spans)
+        public override void AddTags(
+            NormalizedSnapshotSpanCollection spans,
+            SegmentedList<TagSpan<IntraTextAdornmentTag>> adornmentTagSpans)
         {
             try
             {
                 if (spans.Count == 0)
-                    return [];
+                    return;
 
+                // If the snapshot has changed, we can't use any of the cached data, as it is associated with the
+                // original snapshot they were created against.
                 var snapshot = spans[0].Snapshot;
-                if (snapshot != _cacheSnapshot)
-                {
-                    // Calculate UI elements
-                    _cache.Clear();
-                    _cacheSnapshot = snapshot;
-
-                    // Calling into the InlineParameterNameHintsDataTaggerProvider which only responds with the current
-                    // active view and disregards and requests for tags not in that view
-                    var fullSpan = new SnapshotSpan(snapshot, 0, snapshot.Length);
-                    var tags = _tagAggregator.GetTags(new NormalizedSnapshotSpanCollection(fullSpan));
-                    foreach (var tag in tags)
-                    {
-                        // Gets the associated span from the snapshot span and creates the IntraTextAdornmentTag from the data
-                        // tags. Only dealing with the dataTagSpans if the count is 1 because we do not see a multi-buffer case
-                        // occurring
-                        var dataTagSpans = tag.Span.GetSpans(snapshot);
-                        if (dataTagSpans.Count == 1)
-                        {
-                            _cache.Add((tag, tagSpan: null));
-                        }
-                    }
-                }
 
                 var document = snapshot.GetOpenDocumentInCurrentContextWithChanges();
-                var classify = document != null && _taggerProvider.EditorOptionsService.GlobalOptions.GetOption(InlineHintsViewOptionsStorage.ColorHints, document.Project.Language);
+                if (document is null)
+                    return;
+
+                var colorHints = _taggerProvider.EditorOptionsService.GlobalOptions.GetOption(InlineHintsViewOptionsStorage.ColorHints, document.Project.Language);
+
+                using var _1 = SegmentedListPool.GetPooledList<TagSpan<InlineHintDataTag>>(out var dataTagSpans);
+                _underlyingTagger.AddTags(spans, dataTagSpans);
```

Comment:
new logic is much simpler.  our client asks for tags for a certain span, so we just ask the data tagger for *its* tags from that same span.  no need to cache more or compute more than what we're directly asked for.